src/rules.rs

```diff
@@ -146,193 +147,105 @@ mod tests {
-        let mut mask = TripleMask::default();
-
-        mask = match_type_rule_subject(&subject, mask, &type_map, &rules);
-
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-        assert!(!mask.is_set(&TripleMask::OBJECT));
-    }
-    #[test]
-    // Test the type rule for an object that is in the rules & type map
-    fn is_object_set() {
-        let object = Term::NamedNode(NamedNode {
-            iri: "http://example.org/Alice".to_string(),
-        });
-        let rules = set_type_rule("Person");
-        let mut type_map = HashMap::new();
-        type_map.insert("http://example.org/Alice".to_string(), "Person".to_string());
+    #[rstest]
+    // Subject is in the rules & type index
+    #[case(true, "Alice", "Alice", "Person", "Person", true, false)]
+    // Subject is in the type index, not in the rules
+    #[case(true, "Alice", "Alice", "Person", "Bank", false, false)]
+    // Subject is not in the type index
+    #[case(true, "Alice", "BankName", "Bank", "Bank", false, false)]
+    // Object is in the rules & type index
+    #[case(false, "Alice", "Alice", "Person", "Person", false, true)]
+    fn type_rule(
+        #[case] is_subject: bool,
+        #[case] node_iri: &str,
+        #[case] index_subject: &str,
+        #[case] index_object: &str,
+        #[case] rule_type: &str,
+        #[case] expected_s: bool,
+        #[case] expected_o: bool,
+    ) {
+        let rules = set_type_rule(rule_type);
+        let mut type_index = HashMap::new();
+        type_index.insert(index_subject.to_string(), index_object.to_string());
         let mut mask = TripleMask::default();
-
-        mask = match_type_rule_object(&object, mask, &type_map, &rules);
-
-        assert!(mask.is_set(&TripleMask::OBJECT));
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-    }
-
-    #[test]
-    fn predicate_in_config() {
-        let predicate = NamedNode {
-            iri: "http://example.org/hasName".to_string(),
+        mask = if is_subject {
+            let node = Subject::NamedNode(NamedNode {
+                iri: node_iri.to_string(),
+            });
+            match_type_rule_subject(&node, mask, &type_index, &rules)
+        } else {
+            let node = Term::NamedNode(NamedNode {
+                iri: node_iri.to_string(),
+            });
+            match_type_rule_object(&node, mask, &type_index, &rules)
         };
-        let rules = set_predicate_rule("http://example.org/hasName");
-        let mut mask = TripleMask::default();
-
-        mask = match_predicate_rule(&predicate, mask, &rules);
-
-        assert!(mask.is_set(&TripleMask::OBJECT));
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-    }
-
-    #[test]
-    fn predicate_not_in_config() {
+        assert_eq!(mask.is_set(&TripleMask::SUBJECT), expected_s);
+        assert_eq!(mask.is_set(&TripleMask::OBJECT), expected_o);
+    }
+
+    #[rstest]
+    // Predicate is in the rules
+    #[case("hasName", "hasName", false, true)]
```

Comment:
**suggestion:** don't we expect `expected_s` to always be false here? if so, same suggestion as below